### PR TITLE
[UNR-2763] Add avg/min/max ping measurements

### PR DESCRIPTION
**Pull request by improbable-valentyn**
**Contributions**: We are not currently taking public contributions - see our [contributions](CONTRIBUTING.md) policy. However, we are accepting issues and we do want your [feedback](../README.md#give-us-feedback).

-------

#### Description
SpatialPingComponent can now also report average ping measurements over a specified number of recent pings. You can specify the number of measurements recorded in `PingMeasurementsWindowSize` and get the measurement data by calling `GetAverageData`. There is also a delegate `OnRecordPing` that will be broadcast whenever a new ping measurement is recorded.

#### Tests
Added a UI to display these measurements in LatencyGym:
https://github.com/spatialos/UnrealGDKTestGyms/pull/47

#### Primary reviewers
@m-samiec @improbable-danny

### CODE REVIEW

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/Components/SpatialPingComponent.cpp [hunk 1]
```diff
@@ -181,6 +182,48 @@ void USpatialPingComponent::OnRep_ReplicatedPingID()
...
+	if (LastPingMeasurements.Num() > 0)
+	{
+		double Total = 0.0;
+		for (float Ping : LastPingMeasurements)
+		{
+			Total += Ping;
+		}
+		Data.LastMeasurementsWindowAvg = Total / LastPingMeasurements.Num();
+		Data.LastMeasurementsWindowMin = FMath::Min(LastPingMeasurements);
+		Data.LastMeasurementsWindowMax = FMath::Max(LastPingMeasurements);
```
Comment (improbable-danny) [inline SpatialGDK/Source/SpatialGDK/Private/EngineClasses/Components/SpatialPingComponent.cpp]: This is nice

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/Components/SpatialPingComponent.cpp [hunk 2]
```diff
@@ -181,6 +182,48 @@ void USpatialPingComponent::OnRep_ReplicatedPingID()
...
+
+void USpatialPingComponent::RecordPing(float Ping)
+{
+	LastPingMeasurements.Add(Ping);
+	if (LastPingMeasurements.Num() > PingMeasurementsWindowSize)
+	{
+		LastPingMeasurements.RemoveAt(0);
+	}
+
+	TotalPing += Ping;
```
Comment (improbable-danny) [inline SpatialGDK/Source/SpatialGDK/Private/EngineClasses/Components/SpatialPingComponent.cpp]: Do we need to - LastPingMeasurements[0] before it was removed? Is this the total ping of the sliding window?

Comment (improbable-valentyn) [reply SpatialGDK/Source/SpatialGDK/Private/EngineClasses/Components/SpatialPingComponent.cpp]: No, this is all time total.

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/Components/SpatialPingComponent.cpp [hunk 3]
```diff
@@ -181,6 +182,48 @@ void USpatialPingComponent::OnRep_ReplicatedPingID()
...
+		{
+			Total += Ping;
+		}
+		Data.LastMeasurementsWindowAvg = Total / LastPingMeasurements.Num();
+		Data.LastMeasurementsWindowMin = FMath::Min(LastPingMeasurements);
+		Data.LastMeasurementsWindowMax = FMath::Max(LastPingMeasurements);
+	}
+	Data.WindowSize = LastPingMeasurements.Num();
+
+	Data.TotalAvg = TotalPing / TotalNum;
```
Comment (improbable-danny) [inline SpatialGDK/Source/SpatialGDK/Private/EngineClasses/Components/SpatialPingComponent.cpp]: TotalNum can be 0, converted to float but someone might turn float exceptions on

Comment (improbable-valentyn) [reply SpatialGDK/Source/SpatialGDK/Private/EngineClasses/Components/SpatialPingComponent.cpp]: Good catch, will update!

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/Components/SpatialPingComponent.cpp [hunk 4]
```diff
@@ -181,6 +182,51 @@ void USpatialPingComponent::OnRep_ReplicatedPingID()
 	}
 }
 
+FSpatialPingAverageData USpatialPingComponent::GetAverageData() const
+{
+	FSpatialPingAverageData Data = {};
+
+	if (LastPingMeasurements.Num() > 0)
+	{
+		double Total = 0.0;
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/EngineClasses/Components/SpatialPingComponent.cpp]:
```suggestion
		float Total = 0.0f;
```

File: SpatialGDK/Source/SpatialGDK/Public/EngineClasses/Components/SpatialPingComponent.h [hunk 5]
```diff
@@ -79,4 +117,13 @@ class SPATIALGDK_API USpatialPingComponent : public UActorComponent
 
 	UFUNCTION(Server, Unreliable, WithValidation)
 	virtual void SendServerWorkerPingID(uint16 PingID);
+
+	void RecordPing(float Ping);
+
+	TArray<float> LastPingMeasurements;
+
+	double TotalPing = 0.0;
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Public/EngineClasses/Components/SpatialPingComponent.h]:
```suggestion
	float TotalPing = 0.0f;
```

**Pull request merged**